### ISSUE TITLE
[Feature Request] Multiple Dispatch

### ISSUE THREAD

**Issue by ParadaCarleton** [modularml/mojo#407]
### Review Mojo's priorities

- [X] I have read the [roadmap and priorities](https://docs.modular.com/mojo/roadmap.html#overall-priorities) and I believe this request falls within the priorities.

### What is your request?

```shell
Multiple dispatch in Mojo.
```

### What is your motivation for this change?

```shell
Multiple dispatch is a programming paradigm/feature where runtime polymorphism of functions depends on the types of *all* arguments of a function, not *just* the first function. It is a generalization of both object-oriented programming (which allows for single-polymorphism) and of procedural/functional programming styles. By implementing multiple dispatch, procedures can keep the benefits of dynamic dispatch without having to write code in an object-oriented style, and gain substantial flexibility.
```

**Comment by lattner** [comment]
At this point, we're focused on building out the core system features in Mojo, not looking to introduce general extensions to python.  Let's come back to this after Mojo matures.

Also, while this is a nice feature from (eg) Julia, but depends entirely on its compilation model which doesn't work the same way as Mojo's.

**Comment by ParadaCarleton** [comment]
> At this point, we're focused on building out the core system features in Mojo, not looking to introduce general extensions to python. Let's come back to this after Mojo matures.

I've since edited the post substantially to clarify why I think this needs to be implemented early on. I think multimethods should be considered a core feature in the same way object-oriented programming or single dispatch are. 

Many design decisions are downstream of multimethod support, as shown by PyTorch and Julia, both of which succeeded by building their ecosystems around multiple dispatch from day one.

> Also, while this is a nice feature from (eg) Julia, but depends entirely on its compilation model which doesn't work the same way as Mojo's.

I've edited the OP to reflect how multiple dispatch is possible in ahead-of-time compiled languages, such as C++, or any language using single dispatch.
